src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,492 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                }
+            }
+
+            if (
+              total_size >= _ctx.min_bytes || q_results.has_error
+              || _as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            std::optional<size_t> has_errored_request;
+            if (first_run) {
+                has_errored_request = register_waiters(
+                  std::ranges::iota_view{0ull, _ctx.requests.size()});
+            } else {
+                has_errored_request = register_waiters(requests_map);
+            }
+
+            if (has_errored_request) {
+                auto r_i = has_errored_request.value();
+                results[r_i] = read_result(
+                  error_code::not_leader_for_partition);
+                results[r_i].partition
+                  = _ctx.requests[r_i].ktp().get_partition();
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            co_await _completed_waiter_count.wait();
+
+            if (_as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .total_size = total_size,
+                };
+            }
+
+            first_run = false;
+        }
+    }
+};
+
+/*
+ * A fetch exeutor that relies on notifications from raft to determine when to
+ * query partitions rather than querying all partitions at a set polling
+ * interval like the `parallel_fetch_plan_executor`.
+ */
+class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
+public:
+    nonpolling_fetch_plan_executor()
+      : _last_result_size(ss::smp::count, 0) {}
+
+    /**
+     * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
+     */
+    ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
+        start_worker_aborts(plan);
+
+        // start fetching from a random shard to make sure that we fetch data
+        // from all the partitions even if we reach fetch message size limit
+        const ss::shard_id start_shard_idx = random_generators::get_int(
+          ss::smp::count - 1);
+        for (size_t i = 0; i < ss::smp::count; ++i) {
+            auto shard = (start_shard_idx + i) % ss::smp::count;
+
+            ssx::spawn_with_gate(_workers_gate, [&]() mutable {
+                return log_and_ignore_exceptions(start_shard_fetch_worker(
+                  octx, std::move(plan.fetches_per_shard[shard]), 0));
+            });
+        }
+
+        for (;;) {
+            co_await progress_conditions(octx);
+
+            if (octx.should_stop_fetch()) {
+                // Send abort signal to workers and wait for all workers to end
+                // here.
+                co_await abort_workers();
```

Comment:
It seems like if we are leaving `execute_plan` due to an exception, we won't be aborting workers (see also next comment).